app/models/usage_monitoring/alert.rb

```diff
@@ -0,0 +1,147 @@
+    self.inheritance_column = :alert_type
+
+    STI_MAPPING = {
+      "usage_amount" => "UsageMonitoring::UsageAmountAlert",
+      "billable_metric_usage_amount" => "UsageMonitoring::BillableMetricUsageAmountAlert"
+    }
+
+    CURRENT_USAGE_TYPES = %w[usage_amount billable_metric_usage_amount]
+    BILLABLE_METRIC_TYPES = %w[billable_metric_usage_amount]
+
+    default_scope -> { kept }
+
+    belongs_to :organization
+    belongs_to :billable_metric, optional: true
+
+    has_many :thresholds,
+      foreign_key: :usage_monitoring_alert_id,
+      class_name: "UsageMonitoring::AlertThreshold",
+      dependent: :delete_all
+
+    has_many :triggered_alerts,
+      foreign_key: :usage_monitoring_alert_id,
+      class_name: "UsageMonitoring::TriggeredAlert"
+
+    validate :billable_metric_when_required
+
+    def self.find_sti_class(type_name)
+      STI_MAPPING.fetch(type_name).constantize
+    end
+
+    def self.sti_name
+      STI_MAPPING.invert.fetch(name)
+    end
+
+    def subscription
+      @subscription ||= organization
+        .subscriptions
+        .active
+        .order(started_at: :desc)
+        .find_by(external_id: subscription_external_id)
+    end
+
+    def find_thresholds_crossed(current)
+      crossed = []
+      return crossed if current < previous_value
+      return crossed if current < one_time_thresholds_values.first
+
+      if previous_value < one_time_thresholds_values.last
+        crossed += one_time_thresholds_values.filter { |t| t.between?(previous_value, current) }
+      end
+
+      crossed += find_recurring_thresholds_crossed(
+        previous_value, current, recurring_threshold&.value, one_time_thresholds_values.last
+      )
+
+      crossed.uniq.sort
+    end
+
+    def recurring_threshold
+      @recurring_threshold ||= thresholds.find { |th| th.recurring }
+    end
+
+    def one_time_thresholds_values
+      @one_time_thresholds_values ||= thresholds.all.filter_map { |th| th.value unless th.recurring }.uniq.sort
+    end
+
+    def formatted_crossed_thresholds(crossed_threshold_values)
+      regular_thresholds_values, recurring_thresholds_values = crossed_threshold_values.partition do |v|
+        one_time_thresholds_values.include?(v)
+      end
+
+      formatted_regular_thresholds = thresholds
+        .filter { regular_thresholds_values.include?(it.value) }
+        .map { |t| {code: t.code, value: t.value, recurring: false} }
+
+      formatted_recurring_thresholds = recurring_thresholds_values
+        .map { |v| {code: recurring_threshold&.code, value: v, recurring: true} }
+
+      formatted_regular_thresholds + formatted_recurring_thresholds
+    end
+
+    def find_value(current_metrics)
+      raise NotImplementedError
+    end
+
+    private
+
+    def billable_metric_when_required
+      if billable_metric_id.blank? && BILLABLE_METRIC_TYPES.include?(alert_type)
+        errors.add(:billable_metric_id, "is required for `#{alert_type}` alert type")
```

Comment:
We are usually using error_codes rather than english sentences.
The reason behind this is that it allows API users to map these codes with proper translations / handling on their side